src/Symfony/Component/Lock/Store/MongoDbStore.php

```diff
@@ -143,6 +129,41 @@ public function __construct($mongo, array $options = [], float $initialTtl = 300
         }
     }
 
+    /**
+     * Extract default database and collection from given connection uri and remove collection querystring.
+     */
+    private function skimUri(string $uri): string
+    {
+        if (false === $parsedUrl = parse_url($uri)) {
+            throw new InvalidArgumentException(sprintf('The given MongoDB Connection URI "%s" is invalid.', $mongo));
+        }
+
+        $query = [];
+        if (isset($parsedUrl['query'])) {
+            parse_str($parsedUrl['query'], $query);
+        }
+
+        if (null === ($this->options['collection'] ?? null) && isset($query['collection'])) {
```

Comment:
I know this was already the case before your PR, but the precedence looks wrong to me:
the DSN should have higher precedence over options (I just double-checked, that's how the Cache component does also).

/cc @jderusse maybe we should review all DSN parsing logic to ensure the precedence is consistent everywhere
/cc @Nyholm that's a good case for [the DSN component](https://github.com/symfony/symfony/pull/36999) - formalizing this behavior :)

Reply:
:+1: not only in cache and lock. 

messenger Doctrine:
```php
$configuration += $options + $query + static::DEFAULT_OPTIONS;
```
messenger Redis:
```php
parse_str($parsedUrl['query'], $options);
```
messenger Sqs
```php
 $configuration = ['buffer_size'] = $options['buffer_size'] ?? (int) ($query['buffer_size'] ?? self::DEFAULT_OPTIONS['buffer_size']),
```
messenger amqp
```php
$amqpOptions = array_replace_recursive([], $options, $parsedQuery);
```

Reply:
Bad news, we need to fix this. Up for a PR?

Reply:
@nicolas-grekas In order to change the precedence, I think we should detect `dsn` / `options` where a precedence change would cause an effective option change and instead `trigger_deprecation` for now. Then swap the precedence in `symfony` `6.0`. Do you want me to add this detection / `trigger_deprecation` in this PR for `Lock\Store\MongoDbStore` only or should we make a list of ALL `dsn` / `options` precedence affected classes and do them in a separate PR?

Reply:
I've added the precedence change since you guys seem to be classing this as a bug rather than a deprecation in: https://github.com/symfony/symfony/pull/37268

Reply:
There is no way to deprecate this behavior properly. Also this will very unlikely hit anyone. This precedence issue lessens the capabilities of the ops side of things. Basically, having the options take over the DSN means that the source needs to change (the options) instead of some env vars. That's a management issue at this point, and this change fixes a process issue: the env vars should always win to give control to the ops, which are the ones that know better since they're the ones in touch with the effective infra.

TL;DR, this is a bugfix.